Fix `PackageID` inside `BTypeSymbol` of User defined `BTableTypes` being incorrectly set as `ballerina/lang.annotations:0.0.0`

Thushara-Piyasekara opened the pull request:
## Purpose
$subject
With this PR, user defined types of following types will have the correct PackageID in their BTypeSymbol,

- BTableType
- BXMLType
- BFutureType
- BStreamType
- BTypeDescType

Fixes #42167

## Approach
Changed the TypeResolver to generate BTypeSymbols for above Types using the PackageIDs of their enclosing packages.

## Samples
> Provide high-level details about the samples related to this feature.

## Remarks
> List any other known issues, related PRs, TODO items, or any other notes related to the PR.

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

nipunayf commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java:
```diff
@@ -1734,7 +1734,7 @@ private BType resolveTypeDesc(BLangTableTypeNode td, ResolverData data) {
         BTypeSymbol typeSymbol = type.tsymbol;
         tableType.tsymbol = Symbols.createTypeSymbol(SymTag.TYPE, Flags.asMask(EnumSet.noneOf(Flag.class)),
-                typeSymbol.name, typeSymbol.originalName, typeSymbol.pkgID,
+                typeSymbol.name, typeSymbol.originalName, data.env.enclPkg.symbol.pkgID,
```
```suggestion
                typeSymbol.name, typeSymbol.originalName, symEnv.enclPkg.symbol.pkgID,
```

When there are multiple usages of `data.env`, as in this case, we have extracted it to `symEnv` and used that instead. Apply this in other areas where it is applicable.

Thushara-Piyasekara replied:
Thank you for the helpful suggestions. Will work on these changes and let you know

Thushara-Piyasekara replied:
Fixed with be8c54dffd2c10afe0dab61cdf0d91f5fb702ff2. Found one more similar case in `BLangStreamType` and applied it there as well.

nipunayf reviewed (commented):
It appears that there are multiple instances of type symbols that don't have the `pkgID` set correctly. Can we have tests for each such case?

Thushara-Piyasekara commented:
> It appears that there were multiple instances of type symbols that didn't have the `pkgID` set correctly. Can we have tests for each such case?

@nipunayf Added test cases in e3fd6420cf999be065716b3a943e7ac557bb8c5b

nipunayf commented on tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/symbol-package-id/Ballerina.toml:
```diff
@@ -0,0 +1,4 @@
+[package]
+org = "sample_package_id"
+name = "symbol_package_id"
+version = "0.1.0"
```
```suggestion
version = "0.1.0"

```

Thushara-Piyasekara replied:
Fixed with be8c54dffd2c10afe0dab61cdf0d91f5fb702ff2.

nipunayf commented on tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/symbol-package-id/main.bal:
```diff
@@ -0,0 +1,37 @@
+type UserStream stream<int>;
+
+// typedesc type
+type UserTypedesc typedesc<int>;
```
```suggestion
type UserTypedesc typedesc<int>;

```

Thushara-Piyasekara replied:
Fixed with be8c54dffd2c10afe0dab61cdf0d91f5fb702ff2.

nipunayf commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,74 @@
+
+/**
+ * Test cases for user defined builtinType type definition packageIDs
+ *
```
```suggestion
 *
 * @since 2201.9.0
```

Thushara-Piyasekara replied:
Fixed with be8c54dffd2c10afe0dab61cdf0d91f5fb702ff2.

nipunayf commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    @Test(dataProvider = "TypeDefPackageID")
+    public void testTypeDefPackageId(int line, int offset, String name,String srcFile, String packageID) {
```
```suggestion
    public void testTypeDefPackageId(int line, int offset, String name, String srcFile, String packageID) {
```

nipunayf commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,74 @@
+
+    @Test(dataProvider = "TypeDefPackageID")
+    public void testTypeDefPackageId(int line, int offset, String name,String srcFile, String packageID) {
+        Optional<Document> srcDocument = getDocument(project, null, srcFile);
```
```suggestion
        Optional<Document> srcDocument = getDocument(project, null, srcFile);
        assertTrue(srcDocument.isPresent()
```

Thushara-Piyasekara replied:
Fixed with be8c54dffd2c10afe0dab61cdf0d91f5fb702ff2.

nipunayf commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,74 @@
+    @Test(dataProvider = "TypeDefPackageID")
+    public void testTypeDefPackageId(int line, int offset, String name,String srcFile, String packageID) {
+        Optional<Document> srcDocument = getDocument(project, null, srcFile);
+        TypeDefinitionSymbol symbol = (TypeDefinitionSymbol) assertBasicsAndGetSymbol(model, srcDocument.get(), line, offset, name, SymbolKind.TYPE_DEFINITION);
```
You can directly cast it in here, and opt out the subsequent casting

```suggestion
        TypeDefinitionSymbol symbol = (BallerinaTypeDefinitionSymbol) assertBasicsAndGetSymbol(model, srcDocument.get(), line, offset, name, SymbolKind.TYPE_DEFINITION);
```

Thushara-Piyasekara replied:
Fixed with be8c54dffd2c10afe0dab61cdf0d91f5fb702ff2.

nipunayf commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,74 @@
+        assertEquals(actualPackageID.toString(), packageID);
+    }
+
+    @DataProvider(name = "TypeDefPackageID")
```
```suggestion
    @DataProvider(name = "TypeDefPackageIDPosProvider")
```

Thushara-Piyasekara replied:
Fixed with be8c54dffd2c10afe0dab61cdf0d91f5fb702ff2.

nipunayf commented on tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/symbol-package-id/Ballerina.toml:
```diff
@@ -0,0 +1,4 @@
+[package]
+org = "sample_package_id"
```
```suggestion
org = "sample_org"
```

nipunayf reviewed (commented):
It appears that your tests are passing despite the changes. It would be better if you could verify that your tests are failing when your changes are

Thushara-Piyasekara commented:
> > It appears that your tests are passing despite the changes. It would be better if you could verify that your tests are failing when your changes are not present.
> 
> @nipunayf Found the issue and fixed it with [c69a20a](https://github.com/ballerina-platform/ballerina-lang/commit/c69a20aaf2449dee2b74d91a4534b4bb50e3ccc9). But seems like the `type UserXml xml;` is giving `ballerina/lang.annotations:0.0.0` as its `PackageID` regardless of our changes. The problem seems to be in compiler side. I will look into it and update here.

Fixed the issue with BXML Type test case with 3367eda69f16aaf7e38a0fed9930162b19b050e4.

prakanth97 commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,80 @@
+                (BallerinaTypeDefinitionSymbol) assertBasicsAndGetSymbol(model, srcDocument.get(), line, offset, name,
+                        SymbolKind.TYPE_DEFINITION);
+        PackageID actualPackageID = ((AbstractTypeSymbol) (symbol).typeDescriptor()).getBType().tsymbol.pkgID;
+
```
```suggestion
```

prakanth97 commented on tests/ballerina-compiler-api-test/src/test/resources/testng.xml:
```diff
@@ -176,6 +176,7 @@
             <class name="io.ballerina.semantic.api.test.symbols.SymbolsInXMLNavigationExprsTest" />
             <class name="io.ballerina.semantic.api.test.symbols.SymbolsWithinExprTest" />
             <class name="io.ballerina.semantic.api.test.symbols.TypeDefSymbolTest" />
+            <class name="io.ballerina.semantic.api.test.symbols.TypeDefPackageIdTest" />
```
```suggestion
            <class name="io.ballerina.semantic.api.test.symbols.TypeDefPackageIdTest" />
            <class name="io.ballerina.semantic.api.test.symbols.TypeDefSymbolTest" />
```

prakanth97 commented on compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeResolver.java:
```diff
@@ -795,7 +795,7 @@ private BType resolveTypedescTypeDesc(BLangConstrainedType td, ResolverData data
         BTypeSymbol typeSymbol = type.tsymbol;
         constrainedType.tsymbol = Symbols.createTypeSymbol(typeSymbol.tag, typeSymbol.flags, typeSymbol.name,
-                typeSymbol.originalName, typeSymbol.pkgID, constrainedType, typeSymbol.owner,
+                typeSymbol.originalName, data.env.enclPkg.symbol.pkgID, constrainedType, typeSymbol.owner,
```
shall we create a local var 
```
SymbolEnv symEnv = data.env;
 ```
 replace it in other places, seems it used in several places, pls check other places as well.

Thushara-Piyasekara replied:
Fixed in c5ad5f9dc0d36e45baa2a28b1a138d3104c18d4a. Applied the change for other places as well.

prakanth97 commented:
> > It appears that your tests are passing despite the changes. It would be better if you could verify that your tests are failing when your changes are not present.
> 
> @nipunayf Found the issue and fixed it with [c69a20a](https://github.com/ballerina-platform/ballerina-lang/commit/c69a20aaf2449dee2b74d91a4534b4bb50e3ccc9). But seems like the `type UserXml xml;` is giving `ballerina/lang.annotations:0.0.0` as its `PackageID` regardless of our changes. The problem seems to be in compiler side. I will look into it and update here.

As we discussed in the offline. The behaviour is correct for the xml

> > It appears that your tests are passing despite the changes. It would be better if you could verify that your tests are failing when your changes are not present.
> 
> @nipunayf Found the issue and fixed it with [c69a20a](https://github.com/ballerina-platform/ballerina-lang/commit/c69a20aaf2449dee2b74d91a4534b4bb50e3ccc9). But seems like the `type UserXml xml;` is giving `ballerina/lang.annotations:0.0.0` as its `PackageID` regardless of our changes. The problem seems to be in compiler side. I will look into it and update here.

As we discussed in the offline, the behaviour for the `type XML xml` is correct since type node is build-in type hence it us belongs to the `annotation` pkg. If you check the pkgID of the type definition symbol, it will give expected result. 

In the tests, shall we validate the symbol pkgID as well?

prakanth97 commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,80 @@
+import static org.testng.Assert.assertTrue;
+
+/**
+ * Test cases for user defined builtinType type definition packageIDs
```
```suggestion
 * Test cases for user defined builtinType type definition packageIDs.
```
Build failed due to this.

SasinduDilshara commented on tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/symbol-package-id/main.bal:
```diff
@@ -0,0 +1,37 @@
+type UserStream stream<int>;
+
+// typedesc type
+type UserTypedesc typedesc<int>;
```
Shall we add a test with structural types as well

ex:- `type UserTypedesc typedesc<Employee>;`

Thushara-Piyasekara replied:
Added in c5ad5f9dc0d36e45baa2a28b1a138d3104c18d4a.

Thushara-Piyasekara commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,80 @@
+        BallerinaTypeDefinitionSymbol symbol =
+                (BallerinaTypeDefinitionSymbol) assertBasicsAndGetSymbol(model, srcDocument.get(), line, offset, name,
+                        SymbolKind.TYPE_DEFINITION);
+        PackageID actualPackageID = ((AbstractTypeSymbol) (symbol).typeDescriptor()).getBType().tsymbol.pkgID;
```
Added the assertion in c5ad5f9dc0d36e45baa2a28b1a138d3104c18d4a.

Thushara-Piyasekara commented:
> > > It appears that your tests are passing despite the changes. It would be better if you could verify that your tests are failing when your changes are not present.
> > 
> > 
> > @nipunayf Found the issue and fixed it with [c69a20a](https://github.com/ballerina-platform/ballerina-lang/commit/c69a20aaf2449dee2b74d91a4534b4bb50e3ccc9). But seems like the `type UserXml xml;` is giving `ballerina/lang.annotations:0.0.0` as its `PackageID` regardless of our changes. The problem seems to be in compiler side. I will look into it and update here.
> 
> As we discussed in the offline, the behaviour for the `type XML xml` is correct since type node is build-in type hence it us belongs to the `annotation` pkg. If you check the pkgID of the type definition symbol, it will give expected result.
> 
> In the tests, shall we validate the symbol pkgID as well?

Updated the test accordingly in c5ad5f9dc0d36e45baa2a28b1a138d3104c18d4a.

prakanth97 commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,161 @@
+import static org.testng.Assert.assertTrue;
+
+/**
+ * Test cases for user defined builtinType type definition packageIDs.
```
```suggestion
 * Test cases for user defined type packageIDs.
```
WDYT?

Thushara-Piyasekara replied:
I also think its better. Changed in 674b30fdfb428de902755274b12716fab968fe37.

prakanth97 commented on tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/symbol-package-id/main.bal:
```diff
@@ -0,0 +1,72 @@
+// under the License.
+
+// Simple types
+type UserNil null;
```
```suggestion
type UserNil ();
```

Thushara-Piyasekara replied:
Fixed in f20d0cbd1e0c6cd4d6b5714b31f5d81e88bf4fde.

prakanth97 commented on tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/symbol-package-id/main.bal:
```diff
@@ -0,0 +1,72 @@
+// Sequence types
+type UserString string;
+type UserXml_1 xml;
+type UserXml_2 xml<xml:Element>;
```
Shall we add tests for sub type of xml such as `xml:Element`, `xml:Text`, `xml:ProcessingInstruction`, `xml:Comment`

Thushara-Piyasekara replied:
Added test cases in f20d0cbd1e0c6cd4d6b5714b31f5d81e88bf4fde.

prakanth97 commented on tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/symbol-package-id/main.bal:
```diff
@@ -0,0 +1,72 @@
+type UserReadonly readonly;
+type UserJson json;
+type UserByte byte;
+type UserUnion_1 int|float|null;
```
```suggestion
type UserUnion_1 int|float|();
```

Thushara-Piyasekara replied:
Fixed in 52976acc8b722e505c48db1b95c4370b271d5b7d.

prakanth97 commented on tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/symbol-package-id/main.bal:
```diff
@@ -0,0 +1,72 @@
+type UserAnydata anydata;
+type UserNever never;
+type UserReadonly readonly;
+type UserJson json;
```
Add tests for `json[]` and `map<json>`

Thushara-Piyasekara replied:
Added test cases f20d0cbd1e0c6cd4d6b5714b31f5d81e88bf4fde.

prakanth97 commented on tests/ballerina-compiler-api-test/src/test/resources/test-src/symbols/symbol-package-id/main.bal:
```diff
@@ -0,0 +1,72 @@
+type UserFuture future<int>;
+type UserTypedesc_1 typedesc<int>;
+type UserTypedesc_2 typedesc<UserRecord>;
+type UserStream stream<int>;
```
- Add a test for singleton type such as `type UserSingleton 5`.
- Add tests for built-in subtype of int.

Thushara-Piyasekara replied:
Added test cases in f20d0cbd1e0c6cd4d6b5714b31f5d81e88bf4fde.

Thushara-Piyasekara commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,161 @@
+
+                // Simple types
+                {17, 5, "UserNil", "main.bal", "sample_org/symbol_package_id:0.1.0",
+                        "sample_org/symbol_package_id:0.1.0"},
```
Fixed in f20d0cbd1e0c6cd4d6b5714b31f5d81e88bf4fde.

prakanth97 reviewed (approved):
LGTM

MaryamZi commented on tests/ballerina-compiler-api-test/src/test/java/io/ballerina/semantic/api/test/symbols/TypeDefPackageIdTest.java:
```diff
@@ -0,0 +1,191 @@
+    @DataProvider(name = "TypeDefPackageIDPosProvider")
+    public Object[][] getTypeDefPackageIDPos() {
+        return new Object[][]{
+
```
Can we add some tests referring to builtin types also?

Also will the behaviour now be different for

```ballerina
type T1 future;
```

vs 

```ballerina
type T1 future<any|error>;
```

?

In a way, that's OK I guess. The first is more like an alias.

As discussed with @prakanth97 also though, I think when we have

```
type T some-typedesc;
```

only `T` should have an owner, not `some-typedesc` generally. But we don't have the ideal separation atm to do this.

Thushara-Piyasekara replied:
Added test case for `future` builtin type in d63b2090d7d94931913c1329c4dfc5784e5d5164.

The behavior is different for the given scenario,

```ballerina
type T1 future;
type T2 future<any|error>;
```

For T1, the `PackageID` is still `ballerina/lang.annotations:0.0.0`. But for T2, it is `sample_org/symbol_package_id:0.1.0`. Please let me know if this is not the expected behavior.
